components/file_system/src/rate_limiter.rs

```diff
@@ -1,43 +1,318 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
-use super::{IOOp, IOType};
+use super::{condvar::Condvar, IOOp, IOType, IO_TYPE_VARIANTS};
 
-use std::future::{self, Future};
-use std::sync::{atomic, Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tikv_util::time::Instant;
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+enum IOPriority {
+    Low,
+    High,
+}
+
+fn get_priority(io_type: IOType) -> IOPriority {
+    match io_type {
+        IOType::Flush | IOType::Compaction => IOPriority::Low,
+        _ => IOPriority::High,
+    }
+}
+
+/// Used to calibrate actual bytes through after last reset.
+#[derive(Debug)]
+struct BytesCalibrator {
```

Comment:
Let's make this a trait. The default implementation will consume IO budget on request.